clang/lib/Headers/stddef.h

```diff
@@ -49,7 +48,24 @@
 #define __need_rsize_t
 #endif
 #define __need_wchar_t
+#if !defined(__STDDEF_H) || __has_feature(modules)
+/*
+ * __stddef_null.h is special when building without modules: if __need_NULL is
+ * set, then it will unconditionally redefine NULL. To avoid stepping on client
+ * definitions of NULL, __need_NULL should only be set the first time this
+ * header is included, that is when __STDDEF_H is not defined. However, when
+ * building with modules, this header is a textual header and needs to
+ * unconditionally include __stdef_null.h to support multiple submodules
+ * exporting _Builtin_stddef.null. Take module SM with submodules A and B, whose
+ * headers both include stddef.h When SM.A builds, __STDDEF_H will be defined.
+ * When SM.B builds, the definition from SM.A will leak when building without
+ * local submodule visibility. stddef.h wouldn't include __stddef_null.h, and
+ * SM.B wouldn't import _Builtin_stddef.null, and SM.B's `export *` wouldn't
+ * export NULL as expected. When building with modules, always include
+ * __stddef_null.h so that everything works as expected.
+ */
 #define __need_NULL
+#endif
```

Comment:
This change now looks like a (fairly trivial, but still) regression for local submodule visibility mode. In particular:
```c++
#include <stddef.h>
#undef NULL
#include <stddef.h>
```
is supposed to leave `NULL` undefined, but won't do so any more.

Perhaps we should introduce a feature test macro for "no local submodule visibility" mode ("name leakage between submodules" mode?) so that workarounds for the infelicities of that mode remain scoped to it?

Reply:
I'm pretty sure that will leave `NULL` undefined even in LSV because the second include won't re-bring in any of the pcms but I can add that to the test.

Reply:
(I know that because I was trying to make clang/test/Headers/stddefneeds.cpp work in modules and it simply isn't possible according to some of the reviewers in https://reviews.llvm.org/D159064)

Reply:
Added a "anotherPointer" test to clang/test/Modules/stddef.cpp to illustrate.

Reply:
OK, I see: we want a plain include of `<stddef.h>` to define `NULL` only the first time. In non-modules builds, we get this by checking for `__STDDEF_H`, and in modules builds we get this because `__stddef_null.h` is a modular header. (And checking for `__STDDEF_H` isn't sufficient for a modules but non-LSV build because it might leak from a different submodule.) So it doesn't matter whether we define `__need_NULL` for a modules LSV build; it does the same thing either way.

It looks like we would give different behavior in a modules build for something like:
```
#define __need_NULL
#include <stddef.h>
#undef NULL

#include <stddef.h>
```
... which normally results in `NULL` being defined, but with any flavor of modules, results in `NULL` *not* being defined -- because we build `__stddef_null.h` as a modular header, even though it's really not (it's supposed to redefine `NULL` each time it's included). But that's a pre-existing issue.

In any case, I think I'm convinced this change is correct. Thanks for bearing with me :)

Reply:
> we build __stddef_null.h as a modular header, even though it's really not (it's supposed to redefine NULL each time it's included)

But it _has_ to be a modular header because otherwise all the includers would declare `NULL` in their pcm so we'd have A.NULL, B.NULL, etc. Though I don't think you generally use `NULL` in Swift, if you somehow did (like `func doSomething(pointer: <something>? = NULL)` maybe?) that would cause problems.

Reply:
As a general principle, it seems to me that getting C and C++ semantics right should be the primary concern here, and we shouldn't regress correctness in C or C++ to improve the behavior of Swift interop. Though as it happens, in this case the modules behavior does seem like a reasonable choice for C and C++ -- and really the issue is just that we have a divergence in behavior between modules and non-modules compilations.

I wonder if we could change our behavior for `__need_*` in non-modules builds to match the modules behavior: specifically that you only get each piece once, the first time you ask for it, so if you ask for just `NULL` then ask for everything, you don't get `NULL` the second time. If we think that's good enough for modules, then it seems like it should be good enough in general, and means there's less that can go wrong when turning on / off modules.

Reply:
I agree that would be the more predictable behavior; I'm generally opposed with redefining `NULL` or relying on clang to participate in redefinition schemes. But I don't know how much changing that part would break clients, so I'd rather do it in a separate change since I'm _just_ trying to fix the regression building gcc with clang in this fix.

Reply:
Yeah, sorry, I should have been clear: I still think this change is correct as-is and I'm not suggesting making changes in this pull request.